sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,518 @@
+      this.message = message;
+      this.timestamp = timestamp;
+    }
+
+    public MqttMessage getMessage() {
+      return message;
+    }
+
+    public void setMessage(MqttMessage message) {
+      this.message = message;
+    }
+
+    public Instant getTimestamp() {
+      return timestamp;
+    }
+
+    public void setTimestamp(Instant timestamp) {
+      this.timestamp = timestamp;
+    }
+  }
+
+  private static class UnboundedMqttReader extends UnboundedSource.UnboundedReader<byte[]> {
+
+    private final UnboundedMqttSource source;
+
+    private MqttClient client;
+    private byte[] current;
+    private Instant currentTimestamp;
+    private BlockingQueue<MqttMessageWithTimestamp> queue;
+
+    private UnboundedMqttReader(UnboundedMqttSource source) {
+      this.source = source;
+      this.current = null;
+      this.queue = new LinkedBlockingQueue<>();
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      LOGGER.debug("Starting MQTT reader");
+      Read spec = source.spec;
+      try {
+        client = spec.mqttConnectionConfiguration().getClient();
+        client.subscribe(spec.mqttConnectionConfiguration().topic());
+        client.setCallback(new MqttCallback() {
+          @Override
+          public void connectionLost(Throwable cause) {
+            LOGGER.warn("MQTT connection lost", cause);
+            try {
+              close();
+            } catch (Exception e) {
+              // nothing to do
+            }
+          }
+
+          @Override
+          public void messageArrived(String topic, MqttMessage message) throws Exception {
+            LOGGER.trace("Message arrived");
+            queue.put(new MqttMessageWithTimestamp(message, Instant.now()));
+          }
+
+          @Override
+          public void deliveryComplete(IMqttDeliveryToken token) {
+            // nothing to do
+          }
+        });
+        return advance();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      LOGGER.debug("Taking from the pending queue ({})", queue.size());
+      try {
+        MqttMessageWithTimestamp message = queue.take();
+        current = message.getMessage().getPayload();
+        currentTimestamp = message.getTimestamp();
+        return true;
+      } catch (InterruptedException e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public void close() throws IOException {
+      LOGGER.debug("Closing MQTT reader");
+      try {
+        if (client != null) {
+          try {
+            client.disconnect();
+          } finally {
+            client.close();
+          }
+        }
+      } catch (MqttException mqttException) {
+        throw new IOException(mqttException);
+      }
+    }
+
+    @Override
+    public Instant getWatermark() {
+      MqttMessageWithTimestamp message = queue.peek();
+      if (message == null) {
```

Comment:
@dpmills can you comment on the consequences of reporting watermark in this way? I'm not sure what the best practices should be.